Libplanet.Explorer/Queries/StateQuery.cs

```diff
@@ -0,0 +1,135 @@
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Security.Cryptography;
+using GraphQL;
+using GraphQL.Types;
+using Libplanet.Action;
+using Libplanet.Assets;
+using Libplanet.Blockchain;
+using Libplanet.Blockchain.Policies;
+using Libplanet.Blocks;
+using Libplanet.Explorer.GraphTypes;
+
+namespace Libplanet.Explorer.Queries;
+
+public class StateQuery<T>
+    : ObjectGraphType<(IBlockChainStates<T> ChainStates, IBlockPolicy<T> Policy)>
+    where T : IAction, new()
+{
+    public StateQuery()
+    {
+        Name = "StateQuery";
+        Field<NonNullGraphType<FungibleAssetValueType>>(
+            "balance",
+            arguments: new QueryArguments(
+                new QueryArgument<NonNullGraphType<AddressType>> { Name = "owner" },
+                new QueryArgument<NonNullGraphType<ByteStringType>> { Name = "currencyHash" },
+                new QueryArgument<NonNullGraphType<IdGraphType>> { Name = "offsetBlockHash" }
+            ),
+            resolve: ResolveBalance
+        );
+        Field<FungibleAssetValueType>(
+            "totalSupply",
+            arguments: new QueryArguments(
+                new QueryArgument<NonNullGraphType<ByteStringType>> { Name = "currencyHash" },
+                new QueryArgument<NonNullGraphType<IdGraphType>> { Name = "offsetBlockHash" }
+            ),
+            resolve: ResolveTotalSupply
+        );
+    }
+
+    private static Currency GetNativeTokenFromHash(
+        IResolveFieldContext<(IBlockChainStates<T> ChainStates, IBlockPolicy<T> Policy)> context,
+        string paramName
+    )
+    {
+        byte[] currencyHashBytes = context.GetArgument<byte[]>(paramName);
+        HashDigest<SHA1> currencyHash;
+        try
+        {
+            currencyHash = new HashDigest<SHA1>(currencyHashBytes);
+        }
+        catch (Exception e)
+        {
+            throw new ExecutionError(paramName + " must consist of 20 bytes.\n" + e.Message, e);
+        }
+
+        IImmutableSet<Currency> nativeTokens = context.Source.Policy.NativeTokens;
+        try
+        {
+            return nativeTokens.First(token => token.Hash.Equals(currencyHash));
```

Comment:
It's trivial too, but I concern performance trade-offs between those two methods.  🤔 

AFAIK `Enumerable.Single()` does *not* immediately terminate even if it finds a satisfied element early, because it needs to ensure if there are no more satisfied items in the sequence.  In theory, it terminates the earliest (even though it would throws an exception) when two satisfied items appear in a row, e.g., `{ true, true, false, false, false }` as it needs to only two items.  However, if the sequence only one satisfied item and it's appeared early, `Enumerable.Single()` superfluously consumes the whole sequence, e.g., `{ true, false, false, false, false }`, even though it returns the satisfied item without any exception.  😞

On the other hand, `Enumerable.First()` always terminates the earliest possible immediately after it finds the first satisfied item.  If it consumes the whole sequence it's the case where the sequence has no satisfied item at all.

Therefore, in general, `Enumerable.Single()` has a performance penalty compared to `First()`.  If we should use `Enumerable.Single()` instead of `First()` it's the case we need to ensure only one item in the input sequence satisfies the predicate.  However, it's not the case here, because the input enumerable is already a set which guarantees contained `Currency` values does not overlap their `Hash`.